server/testconnection/listener.go

```diff
@@ -0,0 +1,38 @@
+package testconnection
+
+import (
+	"sync"
+)
+
+type NotifierFn func(Job)
+
+type Listener struct {
+	subscriptions map[string][]NotifierFn
+	mutex         sync.Mutex
+}
+
+func NewListener() *Listener {
+	return &Listener{
+		subscriptions: make(map[string][]NotifierFn),
+	}
+}
+
+func (m *Listener) Subscribe(jobID string, notifier NotifierFn) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	m.subscriptions[jobID] = append(m.subscriptions[jobID], notifier)
+}
+
+func (m *Listener) Unsubscribe(jobID string) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	delete(m.subscriptions, jobID)
```

Comment:
for each job ID there is a slice of subscribers, implying that you can have multiple subscribers for a single job. When any client calls `Unsubscribe`, since this function is deleting **all** registered subscribers, all subscribers will stop being notified.

It might make more sense to have a single listener per job, and have some logic to avoid subscribing to a job that already has a subscriber. Another, more complex option is to assign an ID for every subscriber func and track that for unsubscriptions, but seems overkill for this usecase

Reply:
This is meant to be a 1-1 so I'll remove the string slice for the moment

Reply:
make sense